app/controllers/community_memberships_controller.rb

```diff
@@ -9,102 +9,210 @@ class CommunityMembershipsController < ApplicationController
   skip_filter :ensure_consent_given
   skip_filter :ensure_user_belongs_to_community
 
-  def new
-    existing_membership = @current_user.community_memberships.where(:community_id => @current_community.id).first
+  before_filter :ensure_pending_consent, only: [:pending_consent, :consent_given]
 
-    if existing_membership && existing_membership.accepted?
-      flash[:notice] = t("layouts.notifications.you_are_already_member")
-      redirect_to root and return
-    elsif existing_membership && existing_membership.banned?
-      redirect_to access_denied_tribe_memberships_path and return
-    end
+  Form = EntityUtils.define_builder(
+    [:invitation_code, :string],
+    [:email, :string],
+    [:consent, one_of: [nil, "on"]]
+  )
 
-    @skip_terms_checkbox = true if existing_membership && existing_membership.current_terms_accepted?
-    @community_membership = CommunityMembership.new
+  def pending_consent
+    render_pending_consent_form(invitation_code: session[:invitation_code])
   end
 
-  def create
-    # if there already exists one, modify that
-    @community_membership = CommunityMembership.where(person_id: @current_user.id, community_id: @current_community.id).first_or_create
-    request_params = params[:community_membership] || {}
-    @community_membership.update_attributes(request_params
-      .merge({consent: @current_community.consent, status: "pending_email_confirmation"}))
-
-    # if invitation code is stored in session, use it here
-    params[:invitation_code] ||= session[:invitation_code]
-
-    if @current_community.join_with_invite_only? || params[:invitation_code]
-      unless Invitation.code_usable?(params[:invitation_code], @current_community)
-        # abort user creation if invitation is not usable.
-        # (This actually should not happen since the code is checked with javascript)
-        # This could happen if invitation code is coming from hidden field and is wrong/used for some reason
-        session[:invitation_code] = nil # reset code from session if there was issues so that's not used again
-        ApplicationHelper.send_error_notification("Invitation code check did not prevent submiting form, but was detected in the CommunityMembershipsController", "Invitation code error")
-
-        # TODO: if this ever happens, should change the message to something else than "unknown error"
-        flash[:error] = t("layouts.notifications.unknown_error")
-        render :action => :new and return
+  def consent_given
```

Comment:
Should this be `give_consent`? All the other actions that include `create` or `update` are in imperative.